bufferserver/src/main/java/com/datatorrent/bufferserver/server/Server.java

```diff
@@ -545,23 +548,36 @@ public int readSize()
 
   }
 
-  class Subscriber extends AbstractLengthPrependerClient
+  private class Subscriber extends WriteOnlyLengthPrependerClient
   {
     private LogicalNode ln;
 
     Subscriber(LogicalNode ln, int bufferSize)
     {
-      super(1024, bufferSize);
+      super(1024 * 1024, bufferSize == 0 ? 256 * 1024 : bufferSize);
       this.ln = ln;
       ln.addConnection(this);
-      super.write = false;
     }
 
     @Override
-    public void onMessage(byte[] buffer, int offset, int size)
+    public void connected()
     {
-      logger.warn("Received data when no data is expected: {}",
-          Arrays.toString(Arrays.copyOfRange(buffer, offset, offset + size)));
+      super.connected();
+      serverHelperExecutor.submit(new Runnable()
+      {
+        @Override
+        public void run()
+        {
+          final DataList dl = publisherBuffers.get(ln.getUpstream());
+          if (dl != null) {
+            ln.catchUp();
+            dl.addDataListener(ln);
+          } else {
```

Comment:
If a subscriber connects first, it will create DataList and wait for a Publisher to start publishing to it. It is handled by handleSubscriberRequest. In this case, the DataList is deleted explicitly.